gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/store/GemFireContainer.java

```diff
@@ -2001,55 +2007,58 @@ final void localDestroy(final GemFireTransaction tran)
         }
       }
     }
-    else if (this.region == null || this.region.isDestroyed()) {
+    else if (region == null || region.isDestroyed()) {
       final LanguageConnectionContext lcc;
         // [sumedh] A genuine case should be caught at the derby level
         // while other case where it can arise is receiving a GfxdDDLMessage
         // for a region destruction that has already been replayed using
         // the hidden _DDL_STMTS_REGION.
-        if ((this.region == null || ((!Misc.initialDDLReplayInProgress() && tran != null)
+        if ((region == null || ((!Misc.initialDDLReplayInProgress() && tran != null)
             && (lcc = tran.getLanguageConnectionContext()) != null
             && !lcc.isConnectionForRemote()))
             && GemFireXDUtils.TraceDDLReplay) {
 //        if (GemFireXDUtils.TraceDDLReplay) {
           SanityManager.DEBUG_PRINT(GfxdConstants.TRACE_DDLREPLAY,
               "GemFireContainer#destroy: region ["
-                  + (this.region != null ? this.region.getFullPath() : "null")
+                  + (region != null ? region.getFullPath() : "null")
                   + "] for table '" + this.qualifiedName
                   + "' already destroyed");
         throw new RegionDestroyedException(toString(), this.qualifiedName);
       }
     }
-    else if (this.region.isInitialized()) {
+    else if (region.isInitialized()) {
       // before dropping the table, ensure that WAN queues, if any, are
       // drained completely         
-      waitForGatewayQueuesToFlush();      
+      waitForGatewayQueuesToFlush();
+      CacheListener<?, ?> listener = getObjectStoreListener();
+      if (listener != null && !region.isDestroyed()) {
+        region.removeCacheListener(listener);
```

Comment:
Removing explicitly required when thw region is going to be anyways destroyed?

Reply:
Right could be unnecessary. Was just being extra careful that listener should be removed first. Possible that it could lead to deadlock kind of scenarios if it gets removed a bit later because listener also does region delete operation from inside so was not sure what will happen to that if region is in a half destroyed state.

Reply:
ok. thanks.